// express-zod-api/dist/date-in-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZodDateIn = exports.isoDateRegex = void 0;
const zod_1 = require("zod");
const common_helpers_1 = require("./common-helpers");
// simple regex for ISO date, supports the following formats:
// 2021-01-01T00:00:00.000Z
// 2021-01-01T00:00:00Z
// 2021-01-01T00:00:00
// 2021-01-01
exports.isoDateRegex = /^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2}(\.\d{3})?)?Z?$/;
const zodDateInKind = "ZodDateIn";
class ZodDateIn extends zod_1.ZodType {
    _parse(input) {
        const { status, ctx } = this._processInputParams(input);
        if (ctx.parsedType !== zod_1.ZodParsedType.string) {
            (0, zod_1.addIssueToContext)(ctx, {
                code: zod_1.ZodIssueCode.invalid_type,
                expected: zod_1.ZodParsedType.string,
                received: ctx.parsedType,
            });
            return zod_1.INVALID;
        }
        if (!exports.isoDateRegex.test(ctx.data)) {
            (0, zod_1.addIssueToContext)(ctx, {
                code: zod_1.ZodIssueCode.invalid_string,
                validation: "regex",
            });
            status.dirty();
        }
        const date = new Date(ctx.data);
        if (!(0, common_helpers_1.isValidDate)(date)) {
            (0, zod_1.addIssueToContext)(ctx, {
                code: zod_1.ZodIssueCode.invalid_date,
            });
            return zod_1.INVALID;
        }
        return { status: status.value, value: date };
    }
}
exports.ZodDateIn = ZodDateIn;
ZodDateIn.create = () => new ZodDateIn({
    typeName: zodDateInKind,
});
