// express-zod-api/dist/open-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenAPI = void 0;
const openapi3_ts_1 = require("openapi3-ts");
const common_helpers_1 = require("./common-helpers");
const logical_container_1 = require("./logical-container");
const open_api_helpers_1 = require("./open-api-helpers");
const routing_1 = require("./routing");
class OpenAPI extends openapi3_ts_1.OpenApiBuilder {
    ensureUniqSecuritySchemaName(subject) {
        for (const name in this.rootDoc.components?.securitySchemes || {}) {
            if (JSON.stringify(subject) ===
                JSON.stringify(this.rootDoc.components?.securitySchemes?.[name])) {
                return name;
            }
        }
        this.lastSecuritySchemaIds[subject.type] =
            (this.lastSecuritySchemaIds?.[subject.type] || 0) + 1;
        return `${subject.type.toUpperCase()}_${this.lastSecuritySchemaIds[subject.type]}`;
    }
    constructor({ routing, config, title, version, serverUrl, successfulResponseDescription = "Successful response", errorResponseDescription = "Error response", hasSummaryFromDescription = true, }) {
        super();
        this.lastSecuritySchemaIds = {};
        this.addInfo({ title, version }).addServer({ url: serverUrl });
        const endpointCb = (endpoint, path, _method) => {
            const method = _method;
            const commonParams = { path, method, endpoint };
            const [shortDesc, longDesc] = ["short", "long"].map(endpoint.getDescription.bind(endpoint));
            const inputSources = config.inputSources?.[method] || common_helpers_1.defaultInputSources[method];
            const depictedParams = (0, open_api_helpers_1.depictRequestParams)({
                ...commonParams,
                inputSources,
            });
            const operation = {
                responses: {
                    "200": (0, open_api_helpers_1.depictResponse)({
                        ...commonParams,
                        description: successfulResponseDescription,
                        isPositive: true,
                    }),
                    "400": (0, open_api_helpers_1.depictResponse)({
                        ...commonParams,
                        description: errorResponseDescription,
                        isPositive: false,
                    }),
                },
            };
            if (longDesc) {
                operation.description = longDesc;
                if (hasSummaryFromDescription && shortDesc === undefined) {
                    operation.summary = (0, open_api_helpers_1.ensureShortDescription)(longDesc);
                }
            }
            if (shortDesc) {
                operation.summary = (0, open_api_helpers_1.ensureShortDescription)(shortDesc);
            }
            if (endpoint.getTags().length > 0) {
                operation.tags = endpoint.getTags();
            }
            if (depictedParams.length > 0) {
                operation.parameters = depictedParams;
            }
            if (inputSources.includes("body")) {
                operation.requestBody = (0, open_api_helpers_1.depictRequest)(commonParams);
            }
            const securityRefs = (0, open_api_helpers_1.depictSecurityRefs)((0, logical_container_1.mapLogicalContainer)((0, open_api_helpers_1.depictSecurity)(endpoint.getSecurity()), (securitySchema) => {
                const name = this.ensureUniqSecuritySchemaName(securitySchema);
                const scopes = ["oauth2", "openIdConnect"].includes(securitySchema.type)
                    ? endpoint.getScopes()
                    : [];
                this.addSecurityScheme(name, securitySchema);
                return { name, scopes };
            }));
            if (securityRefs.length > 0) {
                operation.security = securityRefs;
            }
            const swaggerCompatiblePath = (0, open_api_helpers_1.reformatParamsInPath)(path);
            this.addPath(swaggerCompatiblePath, {
                ...(this.rootDoc.paths?.[swaggerCompatiblePath] || {}),
                [method]: operation,
            });
        };
        (0, routing_1.routingCycle)({ routing, endpointCb });
        this.rootDoc.tags = config.tags ? (0, open_api_helpers_1.depictTags)(config.tags) : [];
    }
}
exports.OpenAPI = OpenAPI;
