// express-zod-api/dist/result-handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lastResortHandler = exports.defaultResultHandler = exports.createResultHandler = void 0;
const zod_1 = require("zod");
const api_response_1 = require("./api-response");
const common_helpers_1 = require("./common-helpers");
const metadata_1 = require("./metadata");
const createResultHandler = (definition) => definition;
exports.createResultHandler = createResultHandler;
exports.defaultResultHandler = (0, exports.createResultHandler)({
    getPositiveResponse: (output) => {
        const examples = (0, metadata_1.getMeta)(output, "examples") || [];
        const responseSchema = (0, metadata_1.withMeta)(zod_1.z.object({
            status: zod_1.z.literal("success"),
            data: output,
        }));
        for (const example of examples) {
            // forwarding output examples to response schema
            responseSchema.example({
                status: "success",
                data: example,
            });
        }
        return (0, api_response_1.createApiResponse)(responseSchema);
    },
    getNegativeResponse: () => {
        const responseSchema = (0, metadata_1.withMeta)(zod_1.z.object({
            status: zod_1.z.literal("error"),
            error: zod_1.z.object({
                message: zod_1.z.string(),
            }),
        })).example({
            status: "error",
            error: {
                message: (0, common_helpers_1.getMessageFromError)(new Error("Sample error message")),
            },
        });
        return (0, api_response_1.createApiResponse)(responseSchema);
    },
    handler: ({ error, input, output, request, response, logger }) => {
        if (!error) {
            response.status(200).json({
                status: "success",
                data: output,
            });
            return;
        }
        const statusCode = (0, common_helpers_1.getStatusCodeFromError)(error);
        if (statusCode === 500) {
            logger.error(`Internal server error\n${error.stack}\n`, {
                url: request.url,
                payload: input,
            });
        }
        response.status(statusCode).json({
            status: "error",
            error: { message: (0, common_helpers_1.getMessageFromError)(error) },
        });
    },
});
const lastResortHandler = ({ error, logger, response, }) => {
    logger.error(`Result handler failure: ${error.message}.`);
    response
        .status(500)
        .end(`An error occurred while serving the result: ${error.message}.` +
        (error.hasOriginalError()
            ? `\nOriginal error: ${error.getOriginalErrorMessage()}.`
            : ""));
};
exports.lastResortHandler = lastResortHandler;
