// express-zod-api/dist/routing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initRouting = exports.routingCycle = void 0;
const depends_on_method_1 = require("./depends-on-method");
const endpoint_1 = require("./endpoint");
const errors_1 = require("./errors");
const serve_static_1 = require("./serve-static");
const startup_logo_1 = require("./startup-logo");
const routingCycle = ({ routing, endpointCb, staticCb, parentPath, hasCors, }) => {
    Object.entries(routing).forEach(([segment, element]) => {
        segment = segment.trim();
        if (segment.match(/\//)) {
            throw new errors_1.RoutingError("Routing elements should not contain '/' character.\n" +
                `The error caused by ${parentPath
                    ? `'${parentPath}' route that has a '${segment}'`
                    : `'${segment}'`} entry.`);
        }
        const path = `${parentPath || ""}${segment ? `/${segment}` : ""}`;
        if (element instanceof endpoint_1.AbstractEndpoint) {
            const methods = element.getMethods().slice();
            if (hasCors) {
                methods.push("options");
            }
            methods.forEach((method) => {
                endpointCb(element, path, method);
            });
        }
        else if (element instanceof serve_static_1.ServeStatic) {
            if (staticCb) {
                element.apply(path, staticCb);
            }
        }
        else if (element instanceof depends_on_method_1.DependsOnMethod) {
            Object.entries(element.methods).forEach(([method, endpoint]) => {
                endpointCb(endpoint, path, method);
            });
            if (hasCors && Object.keys(element.methods).length > 0) {
                const firstEndpoint = Object.values(element.methods)[0];
                endpointCb(firstEndpoint, path, "options");
            }
        }
        else {
            (0, exports.routingCycle)({
                routing: element,
                endpointCb,
                staticCb,
                hasCors: hasCors,
                parentPath: path,
            });
        }
    });
};
exports.routingCycle = routingCycle;
const initRouting = ({ app, logger, config, routing, }) => {
    if (config.startupLogo !== false) {
        console.log((0, startup_logo_1.getStartupLogo)());
    }
    (0, exports.routingCycle)({
        routing,
        hasCors: !!config.cors,
        endpointCb: (endpoint, path, method) => {
            app[method](path, async (request, response) => {
                logger.info(`${request.method}: ${path}`);
                await endpoint.execute({ request, response, logger, config });
            });
        },
        staticCb: (path, handler) => {
            app.use(path, handler);
        },
    });
};
exports.initRouting = initRouting;
